File: CS.NodeJS.LoadTest.WebApp/webapp/data/dataMigration.js
module.exports = {

	executeAsync: (options) => {

		let latest = options.migrationDAO.getLatestVersion();
		
		return new Promise((res, rej) => {

			options.migrationDAO.getCurrentVersionAsync()
				.then((version) => {

					if (version == latest) {
						console.log("Schema Updated");
						res();
						return;
					}

					//installing schema version
					options.migrationDAO.migrateSchemaVersionAsync(version, latest)
						.then(() => {
							console.log("Migrate Schema Version - OK");

							//installing identity
							options.migrationDAO.migrateIdentityAsync(version, latest)
								.then(() => {
									console.log('Migrate Identity - OK');

									//creating test result
									options.migrationDAO.migrateTestResultAsync(version, latest)
										.then(() => {
											console.log('Migrate Test Result - OK');

											//creating users
											options.migrationDAO.migrateDbmsUsersAsync(version, latest)
												.then(() => {
													console.log('Migrate DBMS Users - OK');
													res();
												})
												.catch(rej);
										})
										.catch(rej);
								})
								.catch(rej);
						})
						.catch(rej);
				})
				.catch(rej);
		});
	}
}

File: CS.DotNetCore.LoadTest/src/CS.DotNetCore.LoadTest.WebApp/Data/DataMigration.cs


namespace CS.DotNetCore.LoadTest.WebApp.Data
{
    using Config;
    using Mongodb;
    using Mysql;
    using Pgsql;
    using System;

    internal class DataMigration : IDisposable
    {
        private IDataMigrationDAO _dataMigrationDAO;
        private bool _disposeDAO;

        internal DataMigration(IDataMigrationDAO migrationDAO)
        {
            _dataMigrationDAO = migrationDAO;
            _disposeDAO = false;
        }

        internal DataMigration(ILoadTestConfig config)
        {
            _disposeDAO = true;

            switch (config.DBMS)
            {
                case LoadTestDBMS.Pgsql:
                    _dataMigrationDAO = new DataMigrationPgsqlDAO(config.PgsqlConnection);
                    break;

                case LoadTestDBMS.Mysql:
                    _dataMigrationDAO = new DataMigrationMysqlDAO(config.MysqlConnection);
                    break;

                case LoadTestDBMS.Mongodb:
                    _dataMigrationDAO = new DataMigrationMongodbDAO(config.MongoConnection);
                    break;

                default:
                    throw new NotImplementedException("DBMS not implemented");
            }
        }

        public void Execute()
        {
            var currentVersion = _dataMigrationDAO.GetCurrentVersion();

            if (currentVersion == _dataMigrationDAO.LatestVersion)
            {
                Console.WriteLine("Schema Updated");
                return;
            }

            //installing latestVersion
            _dataMigrationDAO.MigrateSchemaVersion(currentVersion, _dataMigrationDAO.LatestVersion);
            Console.WriteLine("Migrate Schema Version - OK");

            _dataMigrationDAO.MigrateIdentity(currentVersion, _dataMigrationDAO.LatestVersion);
            Console.WriteLine("Migrate Identity - OK");

            _dataMigrationDAO.MigrateTestResult(currentVersion, _dataMigrationDAO.LatestVersion);
            Console.WriteLine("Migrate Test Results - OK");

            _dataMigrationDAO.MigrateDbmsUsers(currentVersion, _dataMigrationDAO.LatestVersion);
            Console.WriteLine("Migrate DBMS Users - OK");
        }

        public void Dispose()
        {
            if (_disposeDAO)
            {
                _dataMigrationDAO.Dispose();
            }

            _dataMigrationDAO = null;
        }
    }
}
